### PR TITLE
mgr/orch: improve commandline parsing for update_*

### PR DESCRIPTION
**Pull request by jschmid1**
We have to parse options like hostname, network and optional name

Examples:

"myhost"
"myhost=name"
"myhost:1.2.3.4"
"myhost:1.2.3.4=name"
"myhost:1.2.3.0/24"
"myhost:1.2.3.4/24=name"
"myhost:[v2:1.2.3.4:3000]"
"myhost:[v2:1.2.3.4:3000]=name"
"myhost:[v2:1.2.3.4:3000,v1:1.2.3.4:6789]"
"myhost:[v2:1.2.3.4:3000,v1:1.2.3.4:6789]=name"

This patch improves the durability of the code by using
regexes and adds validations for ip_addresses and subnets.

This will now also act as the single parser for all inputs.

Signed-off-by: Joshua Schmid <jschmid@suse.de>

### CODE REVIEW

File: src/pybind/mgr/ssh/module.py [hunk 1]
```diff
@@ -884,7 +884,7 @@ def _create_mon(self, host, network, name):
         # infer whether this is a CIDR network, addrvec, or plain IP
         if '/' in network:
             extra_args = ['--mon-network', network]
-        elif network.startswith('[v') and network.endswith(']'):
```
Comment (jschmid1) [inline src/pybind/mgr/ssh/module.py]: We shouldn't allow '[' or ']' in the syntax as this will mess with shell escaping/regex.

**Comment by liewegas** [comment]
> "myhost:v2:1.2.3.4:3000/24=name"

I don't think this is valid.. it's an v2:-prefixed IP with a CIDR?  Did you mean ``myhost:[v2:1.2.3.4:3000]=name``

> "myhost:v2:1.2.3.4:3000,v1:1.2.3.4:6789/24=name"

I think we should require brackets around any v2: or v1: prefixed "addrvecs"--that's what ceph requires.  e.g., ``myhost:[v2:1.2.3.4:3000,v1:1.2.3.4:6789]=name``

File: src/pybind/mgr/orchestrator.py [hunk 2]
```diff
@@ -28,40 +29,72 @@
...
-      "myhost:1.2.3.4=name" -> ('myhost', '1.2.3.4', 'name')
-      "myhost:1.2.3.0/24" -> ('myhost', '1.2.3.0/24', None)
+      "myhost"
+      "myhost=name"
+      "myhost:1.2.3.4"
+      "myhost:1.2.3.4=name"
+      "myhost:1.2.3.0/24"
+      "myhost:1.2.3.4/24=name"
+      "myhost:v2:1.2.3.4:3000/24=name"
+      "myhost:v2:1.2.3.4:3000,v1:1.2.3.4:6789/24=name"
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]: could you add a test for this in src/pybind/mgr/tests/test_orchestrator.py ?

File: src/pybind/mgr/ssh/module.py (hunk 1 above)
Comment (liewegas) [reply src/pybind/mgr/ssh/module.py]: this is already the case with other ceph commands that take addrvecs.  The user just needs to use single quotes.

**Comment by sebastian-philipp** [comment]
minor nit: I'd prefer if all PRs have `mgr/orchestrator` as prefix. I'm not going to stop anyone merging PRs with a different prefix though.

File: src/pybind/mgr/orchestrator.py (hunk 2 above)
Comment (jschmid1) [reply src/pybind/mgr/orchestrator.py]: will do

**Comment by jschmid1** [comment]
> > "myhost:v2:1.2.3.4:3000/24=name"
> 
> I don't think this is valid.. it's an v2:-prefixed IP with a CIDR? Did you mean `myhost:[v2:1.2.3.4:3000]=name`
> 
> > "myhost:v2:1.2.3.4:3000,v1:1.2.3.4:6789/24=name"
> 
> I think we should require brackets around any v2: or v1: prefixed "addrvecs"--that's what ceph requires. e.g., `myhost:[v2:1.2.3.4:3000,v1:1.2.3.4:6789]=name`

This means that a user has to escape the square brackets when addrvecs are used. We argued about this in the standup and it seems to be acceptable.

I'll also drop the support for providing subnets in the addrvecs case

from `v1:1.2.3.4:6789/24` to `v1:1.2.3.4:6789`

File: src/pybind/mgr/ssh/module.py (hunk 1 above)
Comment (jschmid1) [reply src/pybind/mgr/ssh/module.py]: this needs to be readded when https://github.com/ceph/ceph/pull/31672#issuecomment-554409119 is addressed.

**Comment by jschmid1** [comment]
yikes
``` 
>       from ipaddress import ip_network, ip_address
E       ImportError: No module named ipaddress
```
fixing..

**Comment by tchaikov** [comment]
@jschmid1 as the failure is blocking, i took the liberty to fix it at #31755

**Pull request merged**